### PR TITLE
[MLIR][ONNX] Add OnnxToTorch support for GlobalMaxPool Op

### PR DESCRIPTION
**Pull request by NeverRaR**
https://github.com/nod-ai/SHARK-Turbine/issues/658

### CODE REVIEW

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 1]
```diff
@@ -628,42 +628,39 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
                 });
-  patterns.onOp("Not", 1,
-                [](OpBinder binder, ConversionPatternRewriter &rewriter) {
-                  Torch::ValueTensorType resultType;
-                  Value operand;
-                  if (binder.tensorOperand(operand) ||
-                      binder.tensorResultType(resultType)) {
-                    return failure();
-                  }
+  patterns.onOp(
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:631]: why this change?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -1205,6 +1202,72 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+          return rewriter.notifyMatchFailure(
+              binder.op, "Expected input type having sizes");
+        }
+        ArrayRef<int64_t> inputShape = inputTensorType.getSizes();
+        unsigned inputRank = inputShape.size();
+        if (!resultType || !resultType.hasSizes()) {
+          return rewriter.notifyMatchFailure(
+              binder.op, "Expected result type having sizes");
+        }
+        ArrayRef<int64_t> resultShape = resultType.getSizes();
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1225]: @rsuderman is this one of the situations where we should do a SmallVector rather than an ArrayRef?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 3]
```diff
@@ -628,42 +628,39 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+  patterns.onOp(
+      "Not", 1, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Torch::ValueTensorType resultType;
+        Value operand;
+        if (binder.tensorOperand(operand) ||
+            binder.tensorResultType(resultType)) {
+          return failure();
+        }
 
-                  auto loc = binder.getLoc();
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:640]: ditto. why does the onnx.Not lowering need to be touched at all?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 4]
```diff
@@ -1205,6 +1202,72 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+            binder.getLoc(),
+            Torch::ListType::get(Torch::IntType::get(binder.op->getContext())),
+            outputSize);
+        Value outputTensor;
+        Type returnIndicesType = torch::Torch::ValueTensorType::get(
+            binder.op->getContext(), resultShape,
+            rewriter.getIntegerType(64, true));
+        if (inputRank == 3) {
+          outputTensor = rewriter
+                             .create<Torch::AtenAdaptiveMaxPool1dOp>(
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1249]: use rewriter.replaceOpWithNewOp instead.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 5]
```diff
@@ -1205,6 +1202,72 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        } else if (inputRank == 5) {
+          outputTensor = rewriter
+                             .create<Torch::AtenAdaptiveMaxPool3dOp>(
+                                 binder.getLoc(), resultType, returnIndicesType,
+                                 operand, outputSizeList)
+                             .getResult(0);
+        } else {
+          return failure();
+        }
+        rewriter.replaceOp(binder.op, outputTensor);
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1268]: wouldn't be necessary if we use rewriter.replaceOpWithNewOp instead

**Review by renxida — CHANGES_REQUESTED** [review]

**Comment by renxida** [comment]
Hello! Welcome!

Nice PR & much appreciated! Left some comments for code quality. I hope you find them helpful.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 6]
```diff
@@ -1205,6 +1202,72 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        }
+        Value outputSizeList = rewriter.create<Torch::PrimListConstructOp>(
+            binder.getLoc(),
+            Torch::ListType::get(Torch::IntType::get(binder.op->getContext())),
+            outputSize);
+        Value outputTensor;
+        Type returnIndicesType = torch::Torch::ValueTensorType::get(
+            binder.op->getContext(), resultShape,
+            rewriter.getIntegerType(64, true));
+        if (inputRank == 3) {
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1247]: instead of running through the if statement and then returning, use early exits and just return failure at the end of the function

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (rsuderman) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1226]: Nah this is fine. `resultType` has an array allocated in it that has the reference returned. You only need to be concerned about using `ArrayRef` in situations like `ArrayRef<int64_t> resultShape = {dim0, dim1};`. This is becase we rely on a list initializer that goes out of memory after the constructor.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 7]
```diff
@@ -7,6 +7,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "mlir/Support/LogicalResult.h"
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:10]: This is probably not required.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 8]
```diff
@@ -628,42 +629,39 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+          auto i1ty = rewriter.getI1Type();
+          auto ty = rewriter.getType<Torch::ValueTensorType>(
+              operandTy.getSizes(), i1ty);
+          auto torchqTy = Torch::getScalarTypeForType(i1ty);
+          Value tyConst = rewriter.create<Torch::ConstantIntOp>(
+              binder.getLoc(), rewriter.getType<Torch::IntType>(),
+              rewriter.getIntegerAttr(rewriter.getIntegerType(64),
+                                      static_cast<int64_t>(torchqTy)));
+          Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+          Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(loc, false);
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:655]: If all of this is done by clang-format, then it's strange that it was not caught by CI before.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]
Hi @NeverRaR, I see many things done wrongly in this lowering. Please correct them.

- First of all, you should not be lowering this op to AdaptiveMaxPooling, instead use "AtenMaxPool<n>dOp". 
- Your lowering returns 2 results while the original Onnx op expects 1 result. 
- The logic for the computation of output shape is also wrong, we can't just assume any dynamic dim to be 1.

**Comment by NeverRaR** [comment]
> Hi @NeverRaR, I see many things done wrongly in this lowering. Please correct them.
> 
> * First of all, you should not be lowering this op to AdaptiveMaxPooling, instead use "AtenMaxPooldOp".
> * Your lowering returns 2 results while the original Onnx op expects 1 result.
> * The logic for the computation of output shape is also wrong, we can't just assume any dynamic dim to be 1.
@vivekkhandelwal1 
*The reason for using AdaptiveMaxPooling is that the current torch-mlir does not support AtenAdaptiveMaxPool1d, and my model happens to require this situation.
*Directly replacing onnx.GlobalMaxPool with AdaptiveMaxPool will still use the result it needs (the first result of AtenAdaptiveMaxPool)
*And why it is  assumed that all dynamic dims are 1. Actually, according to the [onnx specification](https://onnx.ai/onnx/operators/onnx__GlobalMaxPool.html):
>Output data tensor from pooling across the input tensor The output tensor has the same rank as the input The first two dimensions of output shape are the same as the input (N x C), while the other dimensions are all 1

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 8 above)
Comment (NeverRaR) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:655]: It seems that Lint Check has not been run by CI?

**Comment by vivekkhandelwal1** [comment]
> * The reason for using AdaptiveMaxPooling is that the current torch-mlir does not support AtenAdaptiveMaxPool1d, and my model happens to require this situation.

If that's the case, you could have added a lowering for AtenMaxPool1dOp.

> * Directly replacing onnx.GlobalMaxPool with AdaptiveMaxPool will still use the result it needs (the first result of AtenAdaptiveMaxPool)
No, that's not true. You can't replace an op with one result with an op with different number of results. See the crash in the CI for the same.
```
  torch-mlir-opt: /_work/torch-mlir/torch-mlir/externals/llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:1645: virtual void mlir::ConversionPatternRewriter::replaceOp(Operation *, ValueRange): Assertion `op->getNumResults() == newValues.size() && "incorrect # of replacement values"' failed.
```

> * And why it is  assumed that all dynamic dims are 1. Actually, according to the [onnx specification](https://onnx.ai/onnx/operators/onnx__GlobalMaxPool.html):
> 
> > Output data tensor from pooling across the input tensor The output tensor has the same rank as the input The first two dimensions of output shape are the same as the input (N x C), while the other dimensions are all 1

If that's the case then there's no need to put the check, just assign value 1.

**Comment by NeverRaR** [comment]
> > * The reason for using AdaptiveMaxPooling is that the current torch-mlir does not support AtenAdaptiveMaxPool1d, and my model happens to require this situation.
> 
> If that's the case, you could have added a lowering for AtenMaxPool1dOp.
> 
> > * Directly replacing onnx.GlobalMaxPool with AdaptiveMaxPool will still use the result it needs (the first result of AtenAdaptiveMaxPool)
> 
> No, that's not true. You can't replace an op with one result with an op with different number of results. See the crash in the CI for the same.
> 
> ```
>   torch-mlir-opt: /_work/torch-mlir/torch-mlir/externals/llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:1645: virtual void mlir::ConversionPatternRewriter::replaceOp(Operation *, ValueRange): Assertion `op->getNumResults() == newValues.size() && "incorrect # of replacement values"' failed.
> ```
> 
> > * And why it is  assumed that all dynamic dims are 1. Actually, according to the [onnx specification](https://onnx.ai/onnx/operators/onnx__GlobalMaxPool.html):
> > 
> > > Output data tensor from pooling across the input tensor The output tensor has the same rank as the input The first two dimensions of output shape are the same as the input (N x C), while the other dimensions are all 1
> 
> If that's the case then there's no need to put the check, just assign value 1.

Okay, I think I understand now. All modifications have been completed.I will added a lowering for AtenMaxPool1dOp in next pr.

**Comment by NeverRaR** [comment]
@vivekkhandelwal1 Does the failure of CI seem unrelated to my code?

**Comment by vivekkhandelwal1** [comment]
> @vivekkhandelwal1 Does the failure of CI seem unrelated to my code?

Yes

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Comment by vivekkhandelwal1** [comment]
@NeverRaR, you need approval from other reviewers as well to merge this.

**Review by renxida — APPROVED** [review]
